core/src/main/java/org/mskcc/cbio/portal/scripts/ImportGeneData.java

```diff
@@ -72,119 +72,190 @@ public static void importData(File geneFile, String genomeBuild) throws IOExcept
+            addReferenceGenomeGenesToDB(refGenesWithSymbolFromNomenClatureAuthority,
                 refGenesWithoutSymbolFromNomenClatureAuthority);
         }
     }
-        
-        /**
-         * Iterate over the genes found in the given maps and try to add them to the DB.
-         * 
-         * @param genesWithSymbolFromNomenClatureAuthority: genes with official symbol
-         * @param genesWithoutSymbolFromNomenClatureAuthority: genes without official symbol (can happen, some entrez genes
-         * have no official symbol yet, but a temporary symbol).
-         * 
-         * @throws DaoException
-         */
-        private static void addGenesToDB(Map<String, Set<CanonicalGene>> genesWithSymbolFromNomenClatureAuthority,
-                  Map<String, Set<CanonicalGene>> genesWithoutSymbolFromNomenClatureAuthority) throws DaoException {
-                     
-          DaoGeneOptimized daoGene = DaoGeneOptimized.getInstance();
-          int nrExisting = 0;
+
+    /**
+     * Adds the genes parsed from the HGNC file into the database.
+     *
+     * @param geneFile HGNC File with gene information
+     * @throws IOException
+     * @throws DaoException
+     */
+    public static void importHGNCData(File geneFile, String genomeBuild) throws IOException, DaoException {
+        Map<String, Set<CanonicalGene>> canonicalGeneMap = new LinkedHashMap<>();
+        Map<String, Set<ReferenceGenomeGene>> refGeneMap = new LinkedHashMap<>();
+        try (FileReader reader = new FileReader(geneFile)) {
+            BufferedReader buf = new BufferedReader(reader);
+            String line;
+            buf.readLine(); // skip first (header) line
+            ProgressMonitor.incrementCurValue(); // for header line
+            while ((line = buf.readLine()) != null) {
+                ProgressMonitor.incrementCurValue();
+                ConsoleUtil.showProgress();
+                String parts[] = line.split("\t", -1); // include trailing empty strings
+                if (!parts[0].matches("[0-9]+")) {
+                    ProgressMonitor.logWarning("Skipping gene with invalid entrez gene id '" + parts[1] + "'");
+                    continue;
+                }
+                int entrezGeneId = Integer.parseInt(parts[0]);
+                String geneSymbol = parts[1];
+                String cytoband = parts[3];
+                String chr = cytoband.split("[pq;c \\|]")[0]; // various strange characters were found in this column
+                String type = parts[4];
+                String strAliases = parts[5];
+                Set<String> aliases = new HashSet<String>();
+                int referenceGenomeId = DaoReferenceGenome.getReferenceGenomeByBuildName(genomeBuild).getReferenceGenomeId();
+
+                // try to get chr from chromosome column if needed
+                // if cytoband is mitochondria then use chromosome M
+                if (!chr.matches("[0-9XY]{1,2}")) {
```

Comment:
This accepts chromosomes such as "XY", "X3", "0" and other invalid patterns. Also chromosome ranges may differ between species. We should probably have a chromosome validator utility which worked based on the genomeBuild. [a future improvement]